doc/src/devdocs/ast.md

```diff
@@ -763,3 +763,61 @@ Boolean properties:
     * 0x01 << 4 = the syntactic control flow within this method is guaranteed to terminate (`:terminates_locally`)
 
     See the documentation of `Base.@assume_effects` for more details.
+
+
+How to interpret line numbers in a `CodeInfo` object:
+
+```
+struct DebugInfo
+    @noinline
+    def::Union{Method,MethodInstance,Symbol}
+    linetable::Union{Nothing,DebugInfo}
+    edges::SimpleVector{DebugInfo}
+    codelocs::String # compressed data
+end
+mutable struct DebugInfoStream
+    def::Union{Method,MethodInstance,Symbol}
+    linetable::Union{Nothing,DebugInfo}
+    edges::Vector{DebugInfo}
+    firstline::Int32 # the starting line for this block (specified by an index of 0)
+    codelocs::Vector{Int32} # for each statement:
+        # index into linetable (if defined), else a line number (in the file represented by def)
+        # then index into edges
+        # then index into edges[linetable]
+end
+```
+
+
+  * `def` : where this DebugInfo was defined (the Method, MethodInstance, or file scope, for example)
+
+  * `linetable`
+
+    Another debuginfo that this was derived from. If `def` is not a Symbol, then it replaces
```

Comment:
"derived" is roughly the opposite of inlining. Being derived from means it started as the same source and then was optimized to form the new object. It exists because otherwise this structure could not quite represent macros (which need to map multiple sources lines to each statement)